### PR TITLE
feat: Tracking cluster image update progress via Rest

### PR DESCRIPTION
**Pull request by Hardikl**
```
cluster_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",phase="validation",phase_name="Pre-update checks",status="completed"} 1.0
cluster_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",phase="ontap_updates",phase_name="ONTAP updates",status="completed"} 1.0
cluster_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",phase="post_update_checks",phase_name="Post-update checks",status="completed"} 1.0

```

### CODE REVIEW

File: conf/rest/9.6.0/clusterupdate.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
+name:       ClusterUpdate
+query:      api/private/cli/cluster/image/show-update-progress
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply conf/rest/9.6.0/clusterupdate.yaml]: Done

**Review by cgrinds — APPROVED** [review]

**Comment by Hardikl** [comment]
```
harvest % curl -s http://localhost:13001/metrics  | grep cl
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="Cluster/management switch support"} 1
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="Manual checks"} 1
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="NFS mounts"} 1
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="ONTAP API deprecation check"} 1
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="SAN compatibility"} 1
cluster_software_validation_labels{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="SnapMirror Synchronous status"} 1

cluster_software_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-01",phase="ONTAP updates",state="completed"} 1
cluster_software_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-02",phase="ONTAP updates",state="completed"} 1
cluster_software_update_labels{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-01",phase="Post-update checks",state="completed"} 1

cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-download-job",node="umeng-aff300-01",state="completed"} 1
cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-failover-job",node="umeng-aff300-01",state="completed"} 1
cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-giveback-job",node="umeng-aff300-01",state="completed"} 1
cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-download-job",node="umeng-aff300-02",state="completed"} 1
cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-failover-job",node="umeng-aff300-02",state="completed"} 1
cluster_software_status_labels{cluster="umeng-aff300-01-02",datacenter="rest",name="do-giveback-job",node="umeng-aff300-02",state="completed"} 1

```

File: cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
+package clustersoftware
+
+import (
+	"github.com/netapp/harvest/v2/cmd/poller/plugin"
+	"github.com/netapp/harvest/v2/pkg/conf"
+	"github.com/netapp/harvest/v2/pkg/matrix"
+	"github.com/netapp/harvest/v2/pkg/slogx"
+	"github.com/netapp/harvest/v2/pkg/tree/node"
+	"github.com/netapp/harvest/v2/pkg/util"
+	"github.com/tidwall/gjson"
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go]: added

File: cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	if err := c.createValidationMetrics(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ClusterSoftware) Run(dataMap map[string]*matrix.Matrix) ([]*matrix.Matrix, *util.Metadata, error) {
+	data := dataMap[c.Object]
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go]: yep, used global_labels directly

**Comment by Hardikl** [comment]
```
harvest % curl -s http://localhost:13001/metrics  | grep cl
# HELP cluster_software_update Metric for cluster_software
# TYPE cluster_software_update gauge
cluster_software_update{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-01",phase="ONTAP updates",state="completed"} 1
cluster_software_update{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-02",phase="ONTAP updates",state="completed"} 1
cluster_software_update{cluster="umeng-aff300-01-02",datacenter="rest",node="umeng-aff300-01",phase="Post-update checks",state="completed"} 1
# HELP cluster_software_status Metric for cluster_software
# TYPE cluster_software_status gauge
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-giveback-job",node="umeng-aff300-01",state="completed"} 1
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-download-job",node="umeng-aff300-02",state="completed"} 1
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-failover-job",node="umeng-aff300-02",state="completed"} 1
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-giveback-job",node="umeng-aff300-02",state="completed"} 1
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-download-job",node="umeng-aff300-01",state="completed"} 1
cluster_software_status{cluster="umeng-aff300-01-02",datacenter="rest",name="do-failover-job",node="umeng-aff300-01",state="completed"} 1
# HELP cluster_software_validation Metric for cluster_software
# TYPE cluster_software_validation gauge
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="SnapMirror Synchronous status"} 1
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="Cluster/management switch support"} 1
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="Manual checks"} 1
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="NFS mounts"} 1
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="ONTAP API deprecation check"} 1
cluster_software_validation{cluster="umeng-aff300-01-02",datacenter="rest",status="warning",update_check="SAN compatibility"} 1

```

**Comment by Hardikl** [comment]
I would be planning to add one new panel `Cluster Software` having 3 tables which has these new 3 metrics.

File: cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	"github.com/netapp/harvest/v2/pkg/conf"
+	"github.com/netapp/harvest/v2/pkg/matrix"
+	"github.com/netapp/harvest/v2/pkg/slogx"
+	"github.com/netapp/harvest/v2/pkg/tree/node"
+	"github.com/netapp/harvest/v2/pkg/util"
+	"github.com/netapp/harvest/v2/third_party/tidwall/gjson"
+	"log/slog"
+)
+
+const clusterSoftware = "cluster_software"
```
Comment (cgrinds) [inline cmd/collectors/rest/plugins/clustersoftware/clustersoftware.go:14]: statusMatrix does not need to be exported

```go
const (
	clusterSoftware = "cluster_software"
	updateMatrix = "update"
	statusMatrix = "status"
	validationMatrix = "validation"
)
```

**Comment by Hardikl** [comment]
image

**Comment by Hardikl** [comment]
image

image


image

image

**Review by cgrinds — APPROVED** [review]

**Pull request merged**